### PR TITLE
Api 3458 hash refresh tokens

### PR DESCRIPTION
**Pull request by lukemarkwordtlibertyits**
Oauth-Proxy will now hash all Codes and Refresh Tokens. Backwards compatibility for unhashed codes and refresh tokens should be removed after 42 days of this PR being in productions.

### CODE REVIEW

File: oauth-proxy/oauthHandlers/authorizeHandler.js [hunk 1]
```diff
@@ -52,9 +53,9 @@ const authorizeHandler = async (
   try {
     await dynamoClient.saveToDynamo(
       dynamo,
-      state,
+      hashString(state, config.hmac_secret),
       "redirect_uri",
-      client_redirect,
+      hashString(client_redirect, config.hmac_secret),
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/authorizeHandler.js:58]: The redirect URI can't be hashed.

The `redirectHandler` needs this to know where to send a user after arriving from Okta.

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentStateStrategy.js [hunk 2]
```diff
@@ -9,12 +11,12 @@ class SaveDocumentStateStrategy {
   async saveDocumentToDynamo(document, tokens) {
     try {
       if (document.state && document.refresh_token) {
-        let state = document.state.S;
+        let hashedDocument = getHashedDocument(document, tokens, this.config);
         await this.dynamoClient.saveToDynamo(
           this.dynamo,
-          state,
+          hashedDocument.state,
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentStateStrategy.js:17]: I'm also realizing now that we can't hash state.

State is retrieved and included in token responses.

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentStateStrategy.js (hunk 2 above)
Comment (lukemarkwordtlibertyits) [reply oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies/saveDocumentStateStrategy.js:17]: agreed

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/pullDocumentStrategies/pullDocumentByCodeStrategy.js [hunk 3]
```diff
@@ -9,19 +9,31 @@ class PullDocumentByCodeStrategy {
     this.config = config;
   }
   async pullDocumentFromDynamo() {
+    let hashedCode = hashString(this.req.body.code, this.config.hmac_secret);
+    let document = await this.getDocumentDynamo(hashedCode);
+
+    // Backwards compatability.
+    // Remove after 42 Days of PR merge (DATE - Fill out before PR).
```
Comment (dbrown-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/pullDocumentStrategies/pullDocumentByCodeStrategy.js:16]: Just these couple dates that could be added.

**Review by dbrown-va — APPROVED** [review]
Looks good and confirmed working locally.

**Review by crolarlibertyva — APPROVED** [review]
All good for me.

**Review by beaugrantham-va — APPROVED** [review]

**Pull request merged**